[BUG] NPE in SupervisorPlaner

If a sub-agent of a SupervisorPlaner is optional then the output will be null whichi will lead to a NPE:
Caused by: java.lang.NullPointerException: Cannot invoke "java.lang.Object.toString()" because the return value of "dev.langchain4j.agentic.scope.AgentInvocation.output()" is null
	at dev.langchain4j.agentic.supervisor.SupervisorPlanner.nextAction(SupervisorPlanner.java:91)
	at dev.langchain4j.agentic.internal.PlannerBasedInvocationHandler$PlannerLoop.onSubagentInvoked(PlannerBasedInvocationHandler.java:423)
	at dev.langchain4j.agentic.internal.AgentExecutor.internalExecute(AgentExecutor.java:77)
	at dev.langchain4j.agentic.internal.AgentExecutor.execute(AgentExecutor.java:39)
	at dev.langchain4j.agentic.internal.AgentExecutor.execute(AgentExecutor.java:27)
	at dev.langchain4j.agentic.internal.PlannerBasedInvocationHandler$PlannerLoop.loop(PlannerBasedInvocationHandler.java:374)
	at dev.langchain4j.agentic.internal.PlannerBasedInvocationHandler.executeAgentMethod(PlannerBasedInvocationHandler.java:203)
	at dev.langchain4j.agentic.internal.PlannerBasedInvocationHandler.invoke(PlannerBasedInvocationHandler.java:180)

--- a/langchain4j-agentic/src/main/java/dev/langchain4j/agentic/supervisor/SupervisorPlanner.java
+++ b/langchain4j-agentic/src/main/java/dev/langchain4j/agentic/supervisor/SupervisorPlanner.java
@@ -1,32 +1,32 @@
 package dev.langchain4j.agentic.supervisor;
 
+import static java.util.stream.Collectors.toMap;
+
+import dev.langchain4j.agentic.internal.Context;
+import dev.langchain4j.agentic.planner.Action;
+import dev.langchain4j.agentic.planner.AgentArgument;
+import dev.langchain4j.agentic.planner.AgentInstance;
+import dev.langchain4j.agentic.planner.ChatMemoryAccessProvider;
+import dev.langchain4j.agentic.planner.InitPlanningContext;
+import dev.langchain4j.agentic.planner.Planner;
+import dev.langchain4j.agentic.planner.PlanningContext;
+import dev.langchain4j.agentic.scope.AgenticScope;
+import dev.langchain4j.agentic.scope.DefaultAgenticScope;
+import dev.langchain4j.invocation.LangChain4jManaged;
+import dev.langchain4j.memory.chat.ChatMemoryProvider;
+import dev.langchain4j.memory.chat.MessageWindowChatMemory;
+import dev.langchain4j.model.chat.ChatModel;
+import dev.langchain4j.service.AiServices;
+import dev.langchain4j.service.ParameterNameResolver;
+import dev.langchain4j.service.memory.ChatMemoryAccess;
 import java.util.List;
 import java.util.Map;
 import java.util.function.Function;
 import java.util.function.Supplier;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
-import dev.langchain4j.agentic.internal.Context;
-import dev.langchain4j.agentic.planner.Action;
-import dev.langchain4j.invocation.LangChain4jManaged;
-import dev.langchain4j.agentic.planner.AgentArgument;
-import dev.langchain4j.agentic.planner.AgentInstance;
-import dev.langchain4j.agentic.planner.ChatMemoryAccessProvider;
-import dev.langchain4j.agentic.planner.InitPlanningContext;
-import dev.langchain4j.service.ParameterNameResolver;
-import dev.langchain4j.agentic.planner.Planner;
-import dev.langchain4j.agentic.planner.PlanningContext;
-import dev.langchain4j.agentic.scope.AgenticScope;
-import dev.langchain4j.agentic.scope.DefaultAgenticScope;
-import dev.langchain4j.memory.chat.ChatMemoryProvider;
-import dev.langchain4j.memory.chat.MessageWindowChatMemory;
-import dev.langchain4j.model.chat.ChatModel;
-import dev.langchain4j.service.AiServices;
-import dev.langchain4j.service.memory.ChatMemoryAccess;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
-
-import static java.util.stream.Collectors.toMap;
 
 public class SupervisorPlanner implements Planner, ChatMemoryAccessProvider {
 
@@ -58,9 +58,15 @@ public class SupervisorPlanner implements Planner, ChatMemoryAccessProvider {
 
     private String request;
 
-    public SupervisorPlanner(ChatModel chatModel, ChatMemoryProvider chatMemoryProvider, int maxAgentsInvocations,
-                             SupervisorContextStrategy contextStrategy, SupervisorResponseStrategy responseStrategy,
-                             Function<AgenticScope, String> requestGenerator, String outputKey, Function<AgenticScope, Object> output) {
+    public SupervisorPlanner(
+            ChatModel chatModel,
+            ChatMemoryProvider chatMemoryProvider,
+            int maxAgentsInvocations,
+            SupervisorContextStrategy contextStrategy,
+            SupervisorResponseStrategy responseStrategy,
+            Function<AgenticScope, String> requestGenerator,
+            String outputKey,
+            Function<AgenticScope, Object> output) {
         this.chatModel = chatModel;
         this.chatMemoryProvider = chatMemoryProvider;
         this.maxAgentsInvocations = maxAgentsInvocations;
@@ -73,22 +79,27 @@ public class SupervisorPlanner implements Planner, ChatMemoryAccessProvider {
 
     @Override
     public void init(final InitPlanningContext initPlanningContext) {
-        this.agents = initPlanningContext.subagents().stream().collect(toMap(AgentInstance::agentId, Function.identity()));
+        this.agents =
+                initPlanningContext.subagents().stream().collect(toMap(AgentInstance::agentId, Function.identity()));
         this.agentsList = initPlanningContext.subagents().stream()
                 .map(SupervisorPlanner::toCard)
                 .collect(Collectors.joining(", "));
 
-        this.request = requestGenerator != null ? requestGenerator.apply(initPlanningContext.agenticScope()) : initPlanningContext.agenticScope().readState("request", "");
+        this.request = requestGenerator != null
+                ? requestGenerator.apply(initPlanningContext.agenticScope())
+                : initPlanningContext.agenticScope().readState("request", "");
         if (responseStrategy == SupervisorResponseStrategy.SCORED) {
-            this.responseAgent = AiServices.builder(ResponseAgent.class).chatModel(chatModel).build();
+            this.responseAgent =
+                    AiServices.builder(ResponseAgent.class).chatModel(chatModel).build();
         }
     }
 
     @Override
     public Action nextAction(PlanningContext planningContext) {
-        String lastResponse = planningContext.previousAgentInvocation() == null ?
-                "" :
-                planningContext.previousAgentInvocation().output().toString();
+        String lastResponse = planningContext.previousAgentInvocation() == null
+                        || planningContext.previousAgentInvocation().output() == null
+                ? ""
+                : planningContext.previousAgentInvocation().output().toString();
         if (loopCount++ >= maxAgentsInvocations) {
             return doneAction(planningContext.agenticScope(), lastResponse, null);
         }
@@ -112,16 +123,19 @@ public class SupervisorPlanner implements Planner, ChatMemoryAccessProvider {
             return "";
         }
 
-        if (type.isPrimitive() || type.isEnum() || type == String.class ||
-                type == Boolean.class || Number.class.isAssignableFrom(type)) {
+        if (type.isPrimitive()
+                || type.isEnum()
+                || type == String.class
+                || type == Boolean.class
+                || Number.class.isAssignableFrom(type)) {
             return name + ": " + type.getSimpleName();
         }
 
-        String fieldsDescription = type.isRecord() ?
-                Stream.of(type.getDeclaredConstructors()[0].getParameters())
+        String fieldsDescription = type.isRecord()
+                ? Stream.of(type.getDeclaredConstructors()[0].getParameters())
                         .map(p -> argumentDescription(p.getType(), ParameterNameResolver.name(p)))
-                        .collect(Collectors.joining(", ")) :
-                Stream.of(type.getDeclaredFields())
+                        .collect(Collectors.joining(", "))
+                : Stream.of(type.getDeclaredFields())
                         .map(f -> argumentDescription(f.getType(), f.getName()))
                         .collect(Collectors.joining(", "));
 
@@ -133,8 +147,10 @@ public class SupervisorPlanner implements Planner, ChatMemoryAccessProvider {
                 ? SUPERVISOR_CONTEXT_PREFIX + "'" + agenticScope.readState(SUPERVISOR_CONTEXT_KEY, "") + "'."
                 : "";
 
-        AgentInvocation agentInvocation = withAgenticScope(agenticScope,
-                () -> planner(agenticScope).plan(agenticScope.memoryId(), agentsList, request, lastResponse, supervisorContext));
+        AgentInvocation agentInvocation = withAgenticScope(
+                agenticScope,
+                () -> planner(agenticScope)
+                        .plan(agenticScope.memoryId(), agentsList, request, lastResponse, supervisorContext));
         LOG.info("Agent Invocation: {}", agentInvocation);
 
         if (agentInvocation.getAgentName().equalsIgnoreCase("done")) {
@@ -178,11 +194,15 @@ public class SupervisorPlanner implements Planner, ChatMemoryAccessProvider {
         if (agenticScope.hasState(key)) {
             Class<?> argType = agent.arguments().stream()
                     .filter(arg -> arg.name().equals(key))
-                    .findFirst().map(AgentArgument::rawType).orElse(null);
+                    .findFirst()
+                    .map(AgentArgument::rawType)
+                    .orElse(null);
             if (argType != null) {
                 Object existingValue = agenticScope.readState(key);
-                // avoid overwriting a structured state with an unstructured argument generated from supervisor's LLM response
-                return !argType.isAssignableFrom(existingValue.getClass()) || argType.isAssignableFrom(value.getClass());
+                // avoid overwriting a structured state with an unstructured argument generated from supervisor's LLM
+                // response
+                return !argType.isAssignableFrom(existingValue.getClass())
+                        || argType.isAssignableFrom(value.getClass());
             }
         }
         return true;
@@ -213,8 +233,8 @@ public class SupervisorPlanner implements Planner, ChatMemoryAccessProvider {
             case LAST -> lastResponse;
             case SUMMARY -> doneResponse;
             case SCORED -> {
-                ResponseScore score = withAgenticScope(agenticScope,
-                        () -> responseAgent.scoreResponses(request, lastResponse, doneResponse));
+                ResponseScore score = withAgenticScope(
+                        agenticScope, () -> responseAgent.scoreResponses(request, lastResponse, doneResponse));
                 LOG.info("Response scores: {}", score);
                 yield score.getScore2() > score.getScore1() ? doneResponse : lastResponse;
             }

